ckanext/unhcr/controllers/extended_package.py

```
@@ -180,3 +181,49 @@ def publish_microdata(self, id):
             message = 'Dataset "%s" publishing to the Microdata library is not completed for the following reason: "%s"'
             toolkit.h.flash_error(message % (dataset['title'], error))
         toolkit.redirect_to('dataset_edit', id=dataset['name'])
+
+    def request_access(self, id):
+        if toolkit.request.method != 'POST':
+            return toolkit.abort(405, 'Method Not Allowed')
+
+        message = toolkit.request.params.get('message')
+        if not message:
+            return toolkit.abort(400, "'message' is required")
+
+        action_context = {'model': model, 'user': toolkit.c.user}
+        try:
+            dataset = toolkit.get_action('package_show')(action_context, {'id': id})
+        except toolkit.ObjectNotFound:
+            return toolkit.abort(404, 'Dataset not found')
+        except toolkit.NotAuthorized:
+            return toolkit.abort(403, 'Not Authorized')
+
+        if toolkit.h.can_download(dataset):
+            toolkit.h.flash_notice(
+                'You already have access to download resources from {}'.format(
+                    dataset['title']
+                )
+            )
+            return toolkit.redirect_to('dataset_read', id=dataset['id'])
+
+        org_admins = mailer.get_request_access_email_recipients(dataset)
+        if not org_admins:
+            pass
```

I'm not really sure what we should do if we end up here. Should we throw an error? Fall back to emailing sysadmins?

---

475ad81